pump/storage/vlog.go

```diff
@@ -413,6 +424,10 @@ func (vlog *valueLog) scanRequests(start valuePointer, fn func(*request) error)
 //    that every single binlog added would be visited
 // 2. If GC is running concurrently, logFiles may be closed and deleted, thus breaking the scanning.
 func (vlog *valueLog) scan(start valuePointer, fn func(vp valuePointer, record *Record) error) error {
+	vlog.gcGuard <- struct{}{}
```

Comment:
why `scan` and `gcTS` should be mutually exclusive.
if we can guarantee that `start` is safe(which won't be removed), it's fine that `scan` and `gcTS` would be executed simultaneously

Reply:
But it seems we have nothing in the program to make sure the scanned range doesn't overlap with the GC range.

Note that a slow `scan` may be in progress, while `gcTS` is called, in the current implementation, both calls would consider `start` and `ts` safe.

Reply:
could we try to make sure that the scan operate is safe, but not to avoid conflicts.

in another way, `scan` is a time-consuming and frequent operation, however `vlog gc` will be triggered hourly. it seems that it has a high probability of running `scan` when `vlog gc` is triggered. we still can't free up disk effectively.

when `done_gc_tso` too far behind `gc_tso`, once `gc` process could be continued a few weeks, if `scan` is running, unfortunately, `vlog gc` will be triggered after a few weeks.

Reply:
`vlog.scan` is only called in two places:

1. When the storage is starting up
1. When the slow chaser tries to catch up from the last unread binlog

These are not frequent operations.

Reply:
fine, but I still suggest adding a retry mechanism for `vlog gc`

Reply:
Updated